### PR TITLE
Generalize plugin API

### PR DESCRIPTION
**Pull request by flash-gordon**
We should have done it earlier, it's way simpler now.

### CODE REVIEW

File: core/lib/rom/plugin_registry.rb [hunk 1]
```diff
@@ -61,48 +27,48 @@ def initialize
...
-      when :command       then commands.adapter(adapter)
-      when :mapper        then mappers.adapter(adapter)
-      when :relation      then relations.adapter(adapter)
-      when :schema        then schemas.adapter(adapter)
+      when :relations then :relation
+      when :commands then :command
+      when :mappers then :mapper
+      when :schemas then :schema
+      else type
       end
```
Comment (solnic) [inline core/lib/rom/plugin_registry.rb]: @flash-gordon should we maybe use the inflector here?

File: core/lib/rom/plugin_base.rb [hunk 2]
```diff
@@ -19,10 +24,11 @@ class PluginBase
...
-    def initialize(mod, options)
-      @mod      = mod
-      @options  = options
-      @type = options.fetch(:type)
+    def initialize(name, mod, options)
+      @name    = name
+      @mod     = mod
+      @options = options
+      @type    = options.fetch(:type)
     end
```
Comment (solnic) [inline core/lib/rom/plugin_base.rb]: Seems like this could just use `Initializer`

File: core/lib/rom/plugins.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+require 'dry/container'
+require 'rom/plugin'
+
+module ROM
+  # Registry of all known plugin types (command, relation, mapper, etc)
+  #
+  # @api private
+  module Plugins
+    @registry = ::Dry::Container.new
```
Comment (solnic) [inline core/lib/rom/plugins.rb]: why not just `extend Dry::Container::Mixin` and call it a day?

File: core/lib/rom/plugin_registry.rb (hunk 1 above)
Comment (flash-gordon) [reply core/lib/rom/plugin_registry.rb]: I thought about it but I see no point, better remove these aliases in the next version.

File: core/lib/rom/plugins.rb (hunk 3 above)
Comment (flash-gordon) [reply core/lib/rom/plugins.rb]: no objection against it, I just started small to have a better understanding of what I'll need

**Comment by flash-gordon** [comment]
Oh, actually it's not. Even more, I added name to `Plugin`, this means this method makes no sense now. I'll remove it now.

**Comment by solnic** [comment]
@flash-gordon I recommend paying attention to coverage diffs in PRs. I've had situations like that few times and it helps in spotting dead code.

**Pull request merged**